src/primaite/agents/git/git_policy.py

```diff
@@ -0,0 +1,301 @@
+import torch
+import torch.nn.functional as F
+from torch import nn
+from torch_geometric.nn import GATConv
+from torch_geometric.nn.pool import global_max_pool
+from torch.optim import Adam
+from transformers import BertModel, BertTokenizer, BitsAndBytesConfig
+from transformers import AutoModelForCausalLM, AutoTokenizer
+from transformers.models.llama.modeling_llama import LlamaForCausalLM
+from typing import List, Tuple, Dict
+import numpy as np
+
+
+class GraphEmbedding(nn.Module):
+    def __init__(self, in_channels, output_dim, hidden_dim, device: str = "cuda:1"):
+        super().__init__()
+        self.gat = GATConv(in_channels=in_channels, out_channels=hidden_dim)
+        self.linear = nn.Linear(in_features=hidden_dim, out_features=output_dim)
+        self.device = device
+
+    def forward(self, x, edge_index):
+        x = self.gat(x.to(self.device), edge_index.to(self.device))
+        x = F.relu(x)
+        x = self.linear(x)
+        x = F.relu(x)
+        return x
+
+
+class TextEncoder(nn.Module):
+    def __init__(self, model_name: str = "bert-base-uncased", device: str = "cpu"):
+        super().__init__()
+        self.device = device
+        self.tokenizer = BertTokenizer.from_pretrained(model_name)
+        self.model = BertModel.from_pretrained(model_name).to(device)
+
+    def forward(self, graph):
+        # This is suboptimal, but I inherited it from JackS.
+        prompt = f"Node features:\n{graph.x}\n\nEdge Index: {graph.edge_index}"
+        input_ids = self.tokenizer.encode(prompt)
+        outputs = self.model(torch.tensor(input_ids).unsqueeze(0).to(self.device))  # type: ignore
+        return outputs.last_hidden_state[:, 0, :]
+
+
+class AlignmentProjector(nn.Module):
+    def __init__(self, in_features, out_features, hidden_dim, device: str = "cuda:1") -> None:
+        super().__init__()
+        self.linear1 = nn.Linear(in_features=in_features, out_features=hidden_dim)
+        self.linear2 = nn.Linear(in_features=hidden_dim, out_features=out_features)
+        self.device = device
+
+    def forward(self, x):
+        x = self.linear1(x.to(self.device))
+        x = F.relu(x)
+        x = self.linear2(x)
+        return F.relu(x)
```

Comment:
do not have a relu at the end here, this will limit x to be positive but we want to cover the entire space

so linear, relu, linear is good

Reply:
done :)